pygmt/datatypes/grid.py

```diff
@@ -3,7 +3,362 @@
 """
 
 import ctypes as ctp
+from typing import ClassVar
+
+import numpy as np
+import xarray as xr
+
+# Constants for lengths of grid header variables.
+#
+# Note: Ideally we should be able to get these constants from the GMT shared library
+# using the ``lib["GMT_GRID_UNIT_LEN80"]`` syntax, but it causes cyclic import error.
+# So we have to hardcode the values here.
+GMT_GRID_UNIT_LEN80 = 80
+GMT_GRID_TITLE_LEN80 = 80
+GMT_GRID_COMMAND_LEN320 = 320
+GMT_GRID_REMARK_LEN160 = 160
+
+# GMT uses single-precision for grids by default, but can be built to use
+# double-precision. Currently, only single-precision is supported.
+gmt_grdfloat = ctp.c_float
+
+
+def _parse_nameunits(nameunits: str) -> tuple[str, str | None]:
+    """
+    Get long_name and units attributes from x_units/y_units/z_units in grid header.
+
+    In GMT grid header, the x_units/y_units/z_units are strings in the form of
+    ``long_name [units]``, in which both ``long_name`` and ``units`` and standard
+    netCDF attributes defined by CF conventions. The ``[units]`` part is optional.
+
+    This function parses the x_units/y_units/z_units string and get the ``long_name``
+    and ``units`` attributes.
+
+    Parameters
+    ----------
+    nameunits
+        The x_units/y_units/z_units string in grid header.
+
+    Returns
+    -------
+    (long_name, units)
+        Tuple of netCDF attributes 'long_name' and 'units'. 'units' may be None.
+
+    Examples
+    --------
+    >>> _parse_nameunits("longitude [degrees_east]")
+    ('longitude', 'degrees_east')
+    >>> _parse_nameunits("latitude [degrees_north]")
+    ('latitude', 'degrees_north')
+    >>> _parse_nameunits("x")
+    ('x', None)
+    >>> _parse_nameunits("y")
+    ('y', None)
+    >>>
+    """
+    parts = nameunits.split("[")
+    long_name = parts[0].strip()
+    units = parts[1].strip("]").strip() if len(parts) > 1 else None
+    return long_name, units
+
+
+class _GMT_GRID_HEADER(ctp.Structure):  # noqa: N801
+    """
+    GMT grid header structure for holding a grid header.
+    """
```

Comment:
Can we separate this _GMT_GRID_HEADER struct out into a separate file (`datatypes/header.py`?), maybe in another PR? Since this will be used for the `GMT_GRID`, `GMT_IMAGE` and `GMT_CUBE` structs.

Reply:
> > Can we separate this _GMT_GRID_HEADER struct out into a separate file (datatypes/header.py?), maybe in another PR? Since this will be used for the GMT_GRID, GMT_IMAGE and GMT_CUBE structs.
>
> As shown in https://github.com/GenericMappingTools/pygmt/tree/datatypes/gmtimage, we can keep _GMT_GRID_HEADER in grid.py and import it like from pygmt.datatypes.grid import _GMT_GRID_HEADER. Maybe it's OK to keep it in grid.py?

Ok, we can keep it in `datatypes/grid.py`, but let's wrap that in a separate PR.